File: bigbluebutton-server/webapps/oflaDemo/WEB-INF/src/applications/demoservice.rb
# JRuby - style
require 'java'
module RedFive
    include_package "org.springframework.core.io"
    include_package "org.red5.server.webapp.oflaDemo"
end
include_class "org.red5.server.api.Red5"
include_class "java.util.HashMap"

#
# demoservice.rb - a translation into Ruby of the ofla demo application, a red5 example.
#
# @author Paul Gregoire
#
class DemoService < RedFive::DemoServiceImpl

    attr_reader :filesMap
    attr_writer :filesMap

	def initialize
	   puts "Initializing ruby demoservice"
	   super
	   @filesMap = HashMap.new
	end

	def getListOfAvailableFLVs
		puts "Getting the FLV files"
		begin
#		    puts "R5 con local: #{Red5::getConnectionLocal}"
#		    puts "Scope: #{Red5::getConnectionLocal.getScope}"
#		    puts "Root path: #{File.expand_path('/')}"
#		    puts "Current path:  #{File.expand_path('webapps/oflaDemo/')}"
            dirname = File.expand_path('webapps/oflaDemo/streams').to_s
			Dir.open(dirname).entries.grep(/\.flv$/) do |dir|
			    dir.each do |flvName|
      			    fileInfo = HashMap.new
      			    stats = File.stat(dirname+'/'+flvName)
      			    fileInfo["name"] = flvName
      			    fileInfo["lastModified"] = stats.mtime
      			    fileInfo["size"] = stats.size || 0
                    @filesMap[flvName] = fileInfo
                    print 'FLV Name:', flvName
                    print 'Last modified date:', stats.mtime
                    print 'Size:', stats.size || 0
                    print '-------'
                end
            end
		rescue Exception => ex
			puts "Error in getListOfAvailableFLVs #{errorType} \n"
			puts "Exception: #{ex} \n"
			puts caller.join("\n");
		end
		return filesMap
	end

	def formatDate(date)
		return date.strftime("%d/%m/%Y %I:%M:%S")
	end

    def method_missing(m, *args)
      super unless @value.respond_to?(m)
      return @value.send(m, *args)
    end

end

File: bigbluebutton-AppShare-server/webapps/oflaDemo/WEB-INF/src/org/red5/server/webapp/oflaDemo/DemoService.java
package org.red5.server.webapp.oflaDemo;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.red5.server.api.IScope;
import org.red5.server.api.Red5;
import org.springframework.core.io.Resource;

public class DemoService {

	protected static Logger log = LoggerFactory.getLogger(DemoService.class);

	private String formatDate(Date date) {
		SimpleDateFormat formatter;
		String pattern = "dd/MM/yy H:mm:ss";
		Locale locale = new Locale("en", "US");
		formatter = new SimpleDateFormat(pattern, locale);
		return formatter.format(date);
	}

	/**
	 * Getter for property 'listOfAvailableFLVs'.
	 *
	 * @return Value for property 'listOfAvailableFLVs'.
	 */
	public Map getListOfAvailableFLVs() {
		IScope scope = Red5.getConnectionLocal().getScope();
		Map<String, Map> filesMap = new HashMap<String, Map>();
		try {
			log.debug("getting the FLV files");
			Resource[] flvs = scope.getResources("streams/*.flv");
			addToMap(filesMap, flvs);

			Resource[] mp3s = scope.getResources("streams/*.mp3");
			addToMap(filesMap, mp3s);

			
		} catch (IOException e) {
			log.error("{}", e);
		}
		return filesMap;
	}

	private void addToMap(Map<String, Map> filesMap, Resource[] files)
			throws IOException {
		if (files != null) {
			for (Resource flv : files) {
				File file = flv.getFile();
				Date lastModifiedDate = new Date(file.lastModified());
				String lastModified = formatDate(lastModifiedDate);
				String flvName = flv.getFile().getName();
				String flvBytes = Long.toString(file.length());
				if (log.isDebugEnabled()) {
					log.debug("flvName: {}", flvName);
					log.debug("lastModified date: {}", lastModified);
					log.debug("flvBytes: {}", flvBytes);
					log.debug("-------");
				}
				Map<String, Object> fileInfo = new HashMap<String, Object>();
				fileInfo.put("name", flvName);
				fileInfo.put("lastModified", lastModified);
				fileInfo.put("size", flvBytes);
				filesMap.put(flvName, fileInfo);
			}
		}
	}

}
